solutions/devsprint-matheus-reis-2/DeliveryAppChallenge/Service/DeliveryApi.swift

```diff
@@ -7,25 +7,57 @@
 
 import Foundation
 
-struct DeliveryApi {
+enum UrlName {
+    static let restaurantList: String = "home_restaurant_list.json"
+}
 
-    func fetchRestaurants(_ completion: ([String]) -> Void) {
+protocol DeliveryApiProtocol {
+    func fetchRestaurants(_ name: String, _ completion: @escaping ([Restaurant]?) -> Void)
+    func searchAddresses(_ completion: ([String]) -> Void)
+    func fetchRestaurantDetails(_ completion: (String) -> Void)
+    func fetchMenuItem(_ completion: (String) -> Void)
+}
 
-        completion(["Restaurant 1", "Restaurant 2", "Restaurant 3"])
+struct DeliveryApi: DeliveryApiProtocol {
+    
+    func fetchRestaurants(_ name: String, _ completion: @escaping ([Restaurant]?) -> Void) {
+        self.request(name, completion: completion)
     }
-
+    
     func searchAddresses(_ completion: ([String]) -> Void) {
-
-        completion(["Address 1", "Address 2", "Address 3"])
+        
     }
-
+    
     func fetchRestaurantDetails(_ completion: (String) -> Void) {
-
-        completion("Restaurant Details")
+        
     }
-
+    
     func fetchMenuItem(_ completion: (String) -> Void) {
-
-        completion("Menu Item")
+        
+    }
+    
+    private func request<T: Decodable>(_ name: String, completion: @escaping (T?) -> Void){
```

Comment:
Muito bom o uso de `Generics` nessa implementação. Porém, temos um problema com a `completion`. Pensando em testes, da forma como está feito, não é possível tratar erros corretamente. O código permite apenas identificar que um erro ocorreu (quando o argumento recebido pela `completion` for `nil`).

A própria linguagem fornece um tipo chamado `Result`, que é usado em situações de sucesso e falha. você pode ler um pouco mais à respeito [nesse artigo](https://www.hackingwithswift.com/articles/161/how-to-use-result-in-swift) do HackingWithSwift e na [documentação](https://developer.apple.com/documentation/swift/result). Usando esse recurso, é possível criar um `enum` com erros de requisição customizados(como 401, 402, 404, 500 e etc) e tratá-los da maneira correta no seu código. Abaixo, um exemplo de como ficaria sua função usando o `Result` e um enum de erros hipotético:

```swift
private func request<T: Decodable>(_ name: String, completion: @escaping(Result<T,RequestError> -> Void) {
        guard let  url = URL(string: "https://raw.githubusercontent.com/devpass-tech/challenge-mvvm-delivery/main/api/\(name)") else { return }
        let dataTask = URLSession.shared.dataTask(with: url) { data, response, error in

            if let error = error {
                completion(.failure(.custom(error))
                return
            }

            if let data = data {
                let jsonDecodable = JSONDecoder()
                jsonDecodable.keyDecodingStrategy = .convertFromSnakeCase
                do {
                    let repo = try jsonDecodable.decode(T.self, from: data)
                    completion(.success(repo))
                } catch(let error) {
                    completion(.failure(.custom(error))
                }
            }
        }
        dataTask.resume()
}
```

O trecho acima é apenas um exemplo de uso. Poderíamos tratar individualmente cada código de status HTTP, por exemplo, indicando erros de autorização, servidor indisponível e etc.

Reply:
Show Matheus, efetuado a tratativa!